### PR TITLE
CAL-162: KlvProcessor impls no longer add attributes if the type does not match

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Provides the **second** part of a two-part fix to allow certain cases of mpegts video clips to ingest without error.

Adds static utility methods for sanitizing calls to `metacard::setAttribute` to prevent unexpected, deserialized types from KLV data. Relevant implementations of `KlvProcessor` defer to these utility methods instead of setting attributes themselves. 

#### Who is reviewing it?
@glenhein 
@bdeining 
@jrnorth 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jlcsmith
@kcwire

#### How should this be tested?
In conjunction with [DDF PR 1360](https://github.com/codice/ddf/pull/1360); ingest video products, both good ones and some with bad packets (unexpected packet length - not necessarily corrupt data), and ensure the products successfully ingest and appear during search. Since security markings might be ignored, be sure to login to a profile that has access to system high. 

Also, a full build should pass with all necessary components. Pay extra attention to see if anything out of the ordinary has changed as a result of this patch. 

#### Any background context you want to provide?
The Stanag 4609 spec provides security classifications as unsigned bytes, which are translated to Shorts in Java. This violates what the attribute descriptor for SECURITY_CLASSIFICATION is expecting (a String) and causes unchecked casting issues throughout the system. 

#### What are the relevant tickets?
[CAL-162](https://codice.atlassian.net/browse/CAL-162)
_See also:_
[CAL-163](https://codice.atlassian.net/browse/CAL-163)
[DDF-2387](https://codice.atlassian.net/browse/DDF-2387)

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: libs/klv/src/test/java/org/codice/alliance/libs/klv/UtilitiesTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    private MetacardImpl metacard;
+
+    @Before
+    public void setup() throws Exception {
+        metacard = new MetacardImpl();
+    }
+
+    @After
+    public void destroy() throws Exception {
```
Comment (bdeining) [inline libs/klv/src/test/java/org/codice/alliance/libs/klv/UtilitiesTest.java]: Can remove this

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java [hunk 2]
```diff
@@ -25,4 +30,30 @@ public static boolean isEmptyString(Serializable serializable) {
...
     }
+
+    static void safelySetAttribute(Metacard metacard, Attribute attribute) {
+        safelySetAttribute(metacard, attribute.getName(), attribute.getValues());
+    }
+
+    static void safelySetAttribute(Metacard metacard, String attributeName, Serializable value) {
+        if (metacard.getMetacardType()
+                .getAttributeDescriptor(attributeName)
+                .getType()
```
Comment (bdeining) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: I'm worried about null safety here :  what happens if `getAttributeDescriptor` or `getType` returns null?

**Comment by jrnorth** [comment]
Isn't CAL-163 the real issue here? The `KLVProcessor`s know the data types of the fields they're providing mappings for, so shouldn't it be sufficient to double-check our implementation classes to make sure the data types match?

**Comment by glenhein** [comment]
My thinking was that it was safer to have the code always check that the correct type was being supplied.

**Comment by Lambeaux** [comment]
I agree with you both. The draw-back of this fix is we're failing with less noise **and** a side-effect: the attribute will be null on the metacard. 

We need a way to discover / hear about additional issues with the `KLVProcessor`s (if any exist) while still respecting the types in the attribute descriptors. Should we add an error log?

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java (hunk 2 above)
Comment (coyotesqrl) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: `getType()` won't be a problem but `getAttributeDescriptor()` might return a null, depending on how this is invoked. It might never happen if your callers are all *correct*, but you shouldn't count on that.

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java [hunk 3]
```diff
@@ -25,4 +30,30 @@ public static boolean isEmptyString(Serializable serializable) {
...
+                .isAssignableFrom(value.getClass())) {
+            metacard.setAttribute(new AttributeImpl(attributeName, value));
+        }
+    }
+
+    static void safelySetAttribute(Metacard metacard, String attributeName,
+            List<Serializable> valueList) {
+        if (valueList.stream()
+                .allMatch(serializable -> metacard.getMetacardType()
+                        .getAttributeDescriptor(attributeName)
```
Comment (coyotesqrl) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: Same issue here w/r/t `getAttributeDescriptor()` potentially returning a null. Also, you could extract `metacard.getMetacardType.getAttributeDescriptor(attributeName).getType().getBinding()` so you're not reprocessing it for each element of the list. Won't save you much in the long run, but it is still more efficient.

**Comment by Lambeaux** [comment]
Comments addressed. Added an extra test w/r/t the null descriptor.

**Comment by jrnorth** [comment]
@Lambeaux Regarding your comment about how to discover additional issues with the `KlvProcessor`s: we should be able to do that by inspecting the code. If we do that then we should be able to prove that none of these checks will fail.

**Comment by Lambeaux** [comment]
Back-port up: https://github.com/codice/alliance/pull/183

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java [hunk 4]
```diff
@@ -25,4 +31,30 @@ public static boolean isEmptyString(Serializable serializable) {
     public static boolean isNotEmptyString(Serializable serializable) {
         return !isEmptyString(serializable);
     }
+
+    static void safelySetAttribute(Metacard metacard, Attribute attribute) {
+        safelySetAttribute(metacard, attribute.getName(), attribute.getValues());
```
Comment (jlcsmith) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: without knowing all the places this is used, can we really know that metacard and attribute won't be null?

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java [hunk 5]
```diff
@@ -25,4 +31,30 @@ public static boolean isEmptyString(Serializable serializable) {
...
+                .isAssignableFrom(value.getClass())) {
+            metacard.setAttribute(new AttributeImpl(attributeName, value));
+        }
+    }
+
+    static void safelySetAttribute(Metacard metacard, String attributeName,
+            List<Serializable> valueList) {
+        AttributeDescriptor descriptor = metacard.getMetacardType()
+                .getAttributeDescriptor(attributeName);
+        if (descriptor != null && valueList.stream()
```
Comment (jlcsmith) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: are you sure that valueList can't be null?

**Comment by jrnorth** [comment]
Sorry, I should have been clearer in my previous comment. I meant that I believe the real problem can be solved by inspecting our code. If we do that then we can be sure our types match and we won't have to check them at runtime.

**Comment by Lambeaux** [comment]
@jrnorth There will always be cases the slip through the cracks, especially when it comes to byte-parsing. Safely checking attributes before setting them is valuable because, at minimum, we gain insight into the issues faster when writing tests (i.e. the attributes are null when they shouldn't be, why? - oh! for some reason the types don't match) without the risk of blowing up.

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java [hunk 6]
```diff
@@ -25,4 +33,36 @@ public static boolean isEmptyString(Serializable serializable) {
     public static boolean isNotEmptyString(Serializable serializable) {
         return !isEmptyString(serializable);
     }
+
+    static void safelySetAttribute(Metacard metacard, Attribute attribute) {
+        safelySetAttribute(metacard, attribute.getName(), attribute.getValues());
+    }
+
+    static void safelySetAttribute(Metacard metacard, String attributeName, Serializable value) {
+        notNull(metacard, "");
```
Comment (bdeining) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: For these `notNull` statements, it would be beneficial to use add the message describing what parameter was null.  i.e. `notNull(metacard, "Metacard cannot be null.`);

File: libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java [hunk 7]
```diff
@@ -25,4 +33,36 @@ public static boolean isEmptyString(Serializable serializable) {
...
+        if (descriptor != null && descriptor.getType()
+                .getBinding()
+                .isAssignableFrom(value.getClass())) {
+            metacard.setAttribute(new AttributeImpl(attributeName, value));
+        }
+    }
+
+    static void safelySetAttribute(Metacard metacard, String attributeName,
+            List<Serializable> valueList) {
+        notNull(metacard);
```
Comment (bdeining) [inline libs/klv/src/main/java/org/codice/alliance/libs/klv/Utilities.java]: Same comment here
